src/terminal_box.rs

```diff
@@ -406,72 +407,79 @@ where
-                                    let pos_offset = mk_pos_offset!(accu_width, bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dash_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += dash_width + gap_width;
-                                }
+                                let dot = (2.0, Some(bottom_offset));
+                                let gap = (3.0, None);
+                                draw_repeated(&[dot, gap]);
                             }
 
                             if metadata.flags.contains(Flags::UNDERCURL) {
                                 let style_line_height = style_line_height.floor();
                                 let bottom_offset = style_line_height * 1.5;
-
-                                let full_width = self.end_x - self.start_x;
-                                let mut accu_width = 0.0;
-                                let mut dot_width = 1.0f32.min(full_width - accu_width);
-
-                                while accu_width < full_width {
-                                    dot_width = dot_width.min(full_width - accu_width);
-
-                                    let dot_bottom_offset = match accu_width as u32 % 8 {
-                                        3..=5 => bottom_offset + style_line_height,
-                                        2 | 6 => bottom_offset + 2.0 * style_line_height / 3.0,
-                                        1 | 7 => bottom_offset + 1.0 * style_line_height / 3.0,
-                                        _ => bottom_offset,
-                                    };
-
-                                    let pos_offset = mk_pos_offset!(accu_width, dot_bottom_offset);
-                                    let underline_quad =
-                                        mk_quad!(pos_offset, style_line_height, dot_width);
-                                    renderer.fill_quad(underline_quad, line_color);
-                                    accu_width += dot_width;
-                                }
+                                let pattern: [(f32, Option<f32>); 8] =
+                                    array::from_fn(|i| match i {
+                                        3..=5 => (1.0, Some(bottom_offset + style_line_height)),
+                                        2 | 6 => (
+                                            1.0,
+                                            Some(bottom_offset + 2.0 * style_line_height / 3.0),
+                                        ),
+                                        1 | 7 => (
+                                            1.0,
+                                            Some(bottom_offset + 1.0 * style_line_height / 3.0),
+                                        ),
+                                        _ => (1.0, Some(bottom_offset)),
```

Comment:
nit.

Should have mentioned this earlier, but I think it's better to be more explicit here now and use:

``` rust
  0 => (1.0, Some(bottom_offset)),
  _ => unreachable!(),
```